src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/Rfc2898DeriveBytes.cs

```diff
@@ -249,30 +250,45 @@ private void Initialize()
         }
 
         // This function is defined as follows:
-        // Func (S, i) = HMAC(S || i) | HMAC2(S || i) | ... | HMAC(iterations) (S || i) 
+        // Func (S, i) = HMAC(S || i) ^ HMAC2(S || i) ^ ... ^ HMAC(iterations) (S || i) 
         // where i is the block number.
         private byte[] Func()
         {
             byte[] temp = new byte[_salt.Length + sizeof(uint)];
             Buffer.BlockCopy(_salt, 0, temp, 0, _salt.Length);
             Helpers.WriteInt(_block, temp, _salt.Length);
 
-            temp = _hmac.ComputeHash(temp);
-            
-            byte[] ret = temp;
-            for (int i = 2; i <= _iterations; i++)
+            byte[] ui = ArrayPool<byte>.Shared.Rent(_blockSize);
+            try
             {
-                temp = _hmac.ComputeHash(temp);
+                Span<byte> uiSpan = new Span<byte>(ui, 0, _blockSize);
 
-                for (int j = 0; j < _blockSize; j++)
+                if (!_hmac.TryComputeHash(temp, uiSpan, out int bytesWritten) || bytesWritten != _blockSize)
+                    throw new CryptographicException();
+
+                byte[] ret = new byte[_blockSize];
+                uiSpan.CopyTo(ret);
+
+                for (int i = 2; i <= _iterations; i++)
                 {
-                    ret[j] ^= temp[j];
+                    if (!_hmac.TryComputeHash(uiSpan, uiSpan, out bytesWritten) || bytesWritten != _blockSize)
+                        throw new CryptographicException();
+
+                    for (int j = 0; j < _blockSize; j++)
+                    {
+                        ret[j] ^= ui[j];
```

Comment:
Oh, I think this answers my question above... we need to xor each time with the initial hashed data?